### PR TITLE
Add custom `stopping_criteria` and  `logits_processor` to `generate`

### LINKED ISSUE #12118: Passing a custom stopping_criteria list to model.generate() yields a multiple value error for that keyword arg
---
name: "\U0001F41B Bug Report"
about: Submit a bug report to help us improve transformers
title: ''
labels: ''
assignees: ''

---

## Environment info

- `transformers` version: 4.6.1
- Platform: macOS-10.15.5-x86_64-i386-64bit
- Python version: 3.8.8
- PyTorch version (GPU?): 1.18.1 (no)
- Tensorflow version (GPU?): N/A
- Using GPU in script?: no
- Using distributed or parallel set-up in script?: no

### Who can help

- set model_kwargs programmatically: @patrickvonplaten
- set stopping_criteria programmatically: @Narsil

## Information

Model I am using (Bert, XLNet ...): GPT2DoubleHeadsModel (pretrained model: distilgpt2)

The problem arises when using:
* [ ] the official example scripts: (give details below)
* [x] my own modified scripts: (give details below): Any script I write that passes a custom StoppingCriteriaList via the stopping_criteria keyword arg of generation_utils.GenerationMixin.generate() seems to reproduce this issue.

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [x] my own task or dataset: (give details below): a simple personal chatbot harness with a custom newline stopping criterion

## To reproduce

1. Load a trained model using transformer.generation_utils.GenerationMixin
2. Define a custom StoppingCriteria and StoppingCriteriaList
3. Pass the custom StoppingCriteriaList as a keyword arg to model.generate(), e.g. model.generate(...stopping_criteria=my_custom_list...)

The above steps will yield a "got multiple values for keyword argument 'stopping_criteria'" error message.

## Expected behavior

Ideally, there would be no error message, and the stopping_criteria kwarg would be passed through normally.

### PR DESCRIPTION
**Pull request by lvwerra**
# What does this PR do?

This PR are continues the work and discussions from #12219 with a fresh start.

It integrates the custom `stopping_criteria` and `logits_processor` with the following logic: 
- the default `stopping_criteria`/`logits_processor` are created from the arguments and model's config 
- if additional, custom `stopping_criteria` and `logits_processor` are passed to `generate`, they are compared with the default list
   - if there is an overlap between the two lists (e.g. a `MaxLengthCriteria` in both lists) an error is thrown
   - if there is no overlap the two lists are merged 

Fixes #12118

### CODE REVIEW

**Review by Narsil — APPROVED** [review]
Looks good and clean.

Only suggestion would be making it slightly more obvious in the docstring that those arguments are intended for power users.

**Comment by lvwerra** [comment]
@Narsil thanks for your feedback and catching that copy-paste error! I added something to the docstring and fixed the error.

@patrickvonplaten any comments?

File: src/transformers/generation_utils.py [hunk 1]
```diff
@@ -712,15 +713,33 @@ def _get_logits_processor(
...
-            stopping_criteria.append(MaxTimeCriteria(max_time=max_time))
-        return stopping_criteria
+            criteria.append(MaxTimeCriteria(max_time=max_time))
+        criteria = self._merge_criteria_processor_list(criteria, stopping_criteria)
+        return criteria
+
+    def _merge_criteria_processor_list(
+        self, default_list: StoppingCriteriaList, custom_list: StoppingCriteriaList
+    ) -> StoppingCriteriaList:
+        if custom_list is None:
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:733]: How can the custom_list be None? I think it can only be empty no? Wouldn't it be better to do:

```python
if len(custom_list) == 0:
    return default_list
```

File: src/transformers/models/rag/modeling_rag.py [hunk 2]
```diff
@@ -1469,6 +1473,14 @@ def generate(
                 conditioned on the previously generated tokens :obj:`inputs_ids` and the batch ID :obj:`batch_id`. This
                 argument is useful for constrained generation conditioned on the prefix, as described in
                 `Autoregressive Entity Retrieval <https://arxiv.org/abs/2010.00904>`__.
+            logits_processor (:obj:`LogitsProcessorList`, `optional`):
+                 Custom logits processors that complement the default logits processors built from arguments and a
```
Comment (patrickvonplaten) [inline src/transformers/models/rag/modeling_rag.py:1477]: I like the doc-string!

File: src/transformers/generation_utils.py [hunk 3]
```diff
@@ -712,15 +713,33 @@ def _get_logits_processor(
...
+            return default_list
+        for default in default_list:
+            for custom in custom_list:
+                if type(custom) is type(default):
+                    raise ValueError(
+                        f"A custom stopping criteria or logits processor of type {type(custom)} was passed to `generate` which is already created with an argument or model config."
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:739]:
```suggestion
                        f"A custom stopping criteria or logits processor of type {type(custom)} was passed "
                        "to `generate` which is already created with an argument or model config."
```

File: src/transformers/generation_utils.py [hunk 4]
```diff
@@ -712,15 +713,33 @@ def _get_logits_processor(
...
+            criteria.append(MaxLengthCriteria(max_length=max_length))
         if max_time is not None:
-            stopping_criteria.append(MaxTimeCriteria(max_time=max_time))
-        return stopping_criteria
+            criteria.append(MaxTimeCriteria(max_time=max_time))
+        criteria = self._merge_criteria_processor_list(criteria, stopping_criteria)
+        return criteria
+
+    def _merge_criteria_processor_list(
+        self, default_list: StoppingCriteriaList, custom_list: StoppingCriteriaList
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:731]:
```suggestion
        self, default_list: List[Any], custom_list: List[Any]
```

File: src/transformers/generation_utils.py [hunk 5]
```diff
@@ -712,15 +713,33 @@ def _get_logits_processor(
...
+    ) -> StoppingCriteriaList:
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:732]:
```suggestion
    ) -> List[Any]:
```

File: src/transformers/generation_utils.py (hunk 3 above)
Comment (patrickvonplaten) [reply src/transformers/generation_utils.py:739]: Also, since this error won't be easy to solve for the user can we try to give her/him as much guidance as possible? we could a) Figure out if it's a `stopping criteria` or a `logits processor` from the type. And then b) IMO we should display as much information as possible. E.g.
```
object_type = "stopping criteria" if isinstance(custom, StoppingCriteria) else "logits processor"
f"A custom {object_type} of type {type(custom)} with values {custom} has been passed to `generate`, but it has already been created with the values {default}. {default} has been created by passing the corresponding arguments to generate or by the config's default values."

Comment (patrickvonplaten) [reply src/transformers/generation_utils.py:739]: Something like this maybe, there is probably a better way to phrase it

Comment (patrickvonplaten) [reply src/transformers/generation_utils.py:739]: Actually what we could do in a second step is to enforce that the __init__ names of the processors have to correspond 1-to-1 to the inputs of `generate()` and then we can actually pass the arguments that need to set to `None` in generate to allow this. Maybe a bit overkill in a first step, but we could do this if we encounter issues about it

File: src/transformers/generation_utils.py (hunk 4 above)
Comment (Narsil) [reply src/transformers/generation_utils.py:731]: They are not  of type `List`. `Union[LogitsProcessorList, StoppingCriteriaList]` is the correct thing (you can define a new type too to make it less verbose maybe)

```python
GenerateList = Union[LogitsProcessorList, StoppingCriteriaList]

    def _merge(... default_list: GenerateList) -> GenerateList:
       ....
```

File: src/transformers/generation_utils.py (hunk 1 above)
Comment (lvwerra) [reply src/transformers/generation_utils.py:733]: The default in `generate` is `None` but I can change it to an empty list if you think that is more consistent.

Comment (lvwerra) [reply src/transformers/generation_utils.py:733]: An empty `StoppingCriteriaList` and `LogitsProcessorList` to be more precise.

File: src/transformers/generation_utils.py (hunk 4 above)
Comment (lvwerra) [reply src/transformers/generation_utils.py]: I have added `Union[LogitsProcessorList, StoppingCriteriaList]`. Should be more careful with type hints in the future. I chose the verbose option because it is quite readable that way.

File: src/transformers/generation_utils.py (hunk 5 above)
Comment (lvwerra) [reply src/transformers/generation_utils.py]: See above discussion.

File: src/transformers/generation_utils.py (hunk 3 above)
Comment (lvwerra) [reply src/transformers/generation_utils.py]: I refined the error message to include a little bit more detail. Let me know if you think it is clearer now.

File: src/transformers/generation_utils.py [hunk 6]
```diff
@@ -712,15 +714,36 @@ def _get_logits_processor(
...
+        custom_list: Union[LogitsProcessorList, StoppingCriteriaList],
+    ) -> Union[LogitsProcessorList, StoppingCriteriaList]:
+        if len(custom_list) == 0:
+            return default_list
+        for default in default_list:
+            for custom in custom_list:
+                if type(custom) is type(default):
+                    object_type = "stopping criteria" if isinstance(custom, StoppingCriteria) else "logits processor"
+                    raise ValueError(
+                        f"A custom {object_type} of type {type(custom)} with values {custom} has been passed to `generate`, but it has already been created with the values {default}. {default} has been created by passing the corresponding arguments to generate or by the model's config default values. If you just want to change the default values of {object_type} consider passing them as arguments to `generate` instead of using a custom {object_type}."
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:743]: @sgugger would be super happy if we can split this into multiple lines for readability ;-)
```suggestion
                        f"A custom {object_type} of type {type(custom)} with values {custom} has been passed to `generate`, "
                        f"but it has already been created with the values {default}. {default} has been created by passing the "
                        "corresponding arguments to generate or by the model's config default values. "
                        f"If you just want to change the default values of {object_type} consider passing them as arguments "
                        f"to `generate` instead of using a custom {object_type}."
```

**Review by patrickvonplaten — APPROVED** [review]
Great work! Thanks a lot for the feature. 

Feel free to merge whenever :-)
